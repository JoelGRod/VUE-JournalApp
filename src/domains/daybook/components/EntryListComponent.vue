<template>
  <div class="entry-list-container">
    <div class="px-2 pt-2">
      <input type="text" class="form-control" placeholder="Search Entry" v-model="term" />
    </div>

    <div class="entry-scrollarea px-2 pt-2">
      <EntryComponent v-for="item in getEntriesByTerm" :key="item.id"></EntryComponent>
    </div>
  </div>
</template>

<script>
import { defineAsyncComponent } from "vue";
import { mapGetters } from "vuex";

export default {
  data() {
    return {
      term: ''
    }
  },
  computed: {
    ...mapGetters('daybook', { 
      getEntriesByTerm: "getEntriesByTerm"
    })
  },
  components: {
    EntryComponent: defineAsyncComponent(() => import("./EntryComponent")),
  },
};
</script>

<style land="scss" scoped>
.entry-list-container {
  border-right: 1px solid #2c3e50;
  height: calc(100vh - 3rem);
}

.entry-scrollarea {
  height: calc(100vh - 6rem);
  overflow: scroll;
}
</style>