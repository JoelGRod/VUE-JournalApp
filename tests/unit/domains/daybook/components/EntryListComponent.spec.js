import { shallowMount } from "@vue/test-utils";
import { createStore } from "vuex";

import EntryListComponent from "@/domains/daybook/components/EntryListComponent";
import daybookStore from "@/domains/daybook/store/daybook";

import { daybookState } from "../../../mock-data/test-daybook-store";
// import getters from "@/domains/daybook/store/daybook/getters";

describe('Testing: EntryListComponent.vue', () => {

    /*
    * 1ยบ Alternative way to create a store for testing-> this gives more 
    * control BUT it's better to maintain simple like 
    * in daybook-module.spec.js.
    * Basically you are creating a mock of the daybookStore 
    * and creating a store with that mock and not with 
    * the daybookStore module directly
    */
    /*
    const daybookStoreMock = {
        namespaced: true,
        state() {
            return {
                ...daybookState
            }
        },
        getters
    };
    const store = createStore({
        modules: {
            daybook: { ...daybookStoreMock }
        }
    });
    */

    // 2ยบ Better Approach, simple is better
    const createVuexStore = (initialState) =>
    createStore({
        modules: {
            daybook: {
                ...daybookStore,
                state() {
                    return { ...initialState }
                }
            }
        }
    });
    const store = createVuexStore( daybookState );

    const routerMock = {
        push: jest.fn()
    };

    let wrapper;
    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallowMount(EntryListComponent, {
            global: {
                mocks: {
                    $router: routerMock
                },
                plugins: [store]
            }
        });
    });

    test('should match the snapshot', () => {
        // Assert
        expect( wrapper.html() ).toMatchSnapshot();
    });

    test('should call getEntriesByTerm without term and show 2 entries', () => {
        // Arrange
        // console.log( wrapper.html() );
        // Act
        const renderedEntries = wrapper.findAll('entry-component-stub');
        // Assert
        expect( renderedEntries.length ).toBe(2);
    });

    test('should call getEntriesByTerm and filter entries', async () => {
        // Arrange
        const searchInput = wrapper.find('.form-control');
        // Act
        await searchInput.setValue('this is');
        const renderedEntries = wrapper.findAll('entry-component-stub');
        // Assert
        expect( renderedEntries.length ).toBe(1);
    });

    test('should redirect to /new when new entry button is clicked', async () => {
        // Arrange
        const searchButton = wrapper.find('.btn.btn-primary.mx-3');
        // Act
        await searchButton.trigger('click');
        // Assert
        expect( routerMock.push ).toHaveBeenCalled();
        expect( routerMock.push )
            .toHaveBeenCalledWith(
                { name: 'Daybook-Entry', params: { id: 'new' } }
            );
    });
});