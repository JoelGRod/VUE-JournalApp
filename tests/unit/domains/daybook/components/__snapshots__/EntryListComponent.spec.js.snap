// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing: EntryListComponent.vue should match the snapshot 1`] = `
<div class="entry-list-container">
  <div class="px-2 pt-2"><input type="text" class="form-control" placeholder="Search Entry"></div>
  <div class="mt-2 d-flex flex-column"><button class="btn btn-primary mx-3"><i class="fa fa-plus-circle"></i> New Entry </button></div>
  <div class="entry-scrollarea px-2 pt-2">
    <entry-component-stub entry="[object Object]"></entry-component-stub>
    <entry-component-stub entry="[object Object]"></entry-component-stub>
  </div>
</div>
`;
